src/s3/util/s3Tab.ts

```diff
@@ -5,92 +5,97 @@
+                })
+                this.editor = openEditor
+            } else if (openEditor.document.uri.scheme !== uri.scheme) {
+                // there is a read-only tab open, it needs to be focused, then closed
+                await this.focusAndCloseTab()
+                //good to open in given mode
+                this.editor = await this.window.showTextDocument(doc, { preview: false })
+            }
+        } catch (e) {
+            this.window.showErrorMessage('Error opening file ', e)
+            this.editor = undefined
+        }
+    }
 
-        if (!openEditor) {
-            //there wasn't any open, just display it regularly
-            const doc = await workspace.openTextDocument(this.fileUri)
-            this.editor = await this.window.showTextDocument(doc, { preview: false })
-
-            return this.editor
-        } else if (openEditor.document.uri.scheme === 'file') {
-            //there is a tab for this uri open in edit-mode, just shift focus to it by reopening it with the ViewColumn option
-            await this.window.showTextDocument(openEditor.document, {
-                preview: false,
-                viewColumn: openEditor.viewColumn,
-            })
-            this.editor = openEditor
-
-            return this.editor
-        } else if (openEditor.document.uri.scheme === 's3') {
-            // there is a read-only tab open, it needs to be focused, then closed
-            await this.window.showTextDocument(openEditor.document, {
-                preview: false,
-                viewColumn: openEditor.viewColumn,
-            })
-            await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
-            //good to open in edit-mode
-            const doc = await workspace.openTextDocument(this.fileUri)
-            this.editor = await this.window.showTextDocument(doc, { preview: false })
-
-            return this.editor
+    public async focusAndCloseTab(): Promise<void> {
+        const editor = await this.getActiveEditor()
+        if (!editor) {
+            return
         }
+        await this.window.showTextDocument(editor.document, {
+            preview: false,
+            viewColumn: editor.viewColumn,
+        })
+        await vscode.commands.executeCommand('workbench.action.closeActiveEditor')
+    }
 
-        return undefined
+    /**
+     * Uploads current uri back to parent
+     *
+     * @returns true if upload succe]
+     */
+    public async uploadChangesToS3(): Promise<boolean> {
+        const request = {
+            bucketName: this.s3FileNode.bucket.name,
+            key: this.parent.path + this.s3FileNode.name,
+            fileLocation: this.fileUri,
+            fileSizeBytes: this.s3FileNode.file.sizeBytes!,
+            s3Client: this.s3Client,
+            window: this.window,
+        }
+        try {
+            await uploadWithProgress(request)
+        } catch (e) {
+            //error with upload
```

Comment:
DEFINITELY put out a user-facing error toast here. A failure here with no messaging is a significant gap in the user's expectations. General error logging would be great too

Reply:
did after it returns false, put the toast on the caller

Reply:
I'd say that's fine if the caller is handling multiple errors; is that the case?